src/lib/logger.coffee

```diff
@@ -19,30 +27,49 @@ dockerPromise = initialised.then (config) ->
 
 # Queue up any calls to publish logs whilst we wait to be initialised.
 publish = do ->
-	publishQueue = []
+	publishQueue = [[]]
+	messageIndex = 0
+	publishQueueRemainingBytes = MAX_LOG_BYTE_SIZE
+	logsOverflow = false
 
 	initialised.then (config) ->
 		pubnub = PUBNUB.init(config.pubnub)
 		channel = config.channel
-
-		# Redefine original function
-		publish = (message) ->
-			# Disable sending logs for bandwidth control
-			return if disableLogs
-			if _.isString(message)
-				message = { message }
-
-			_.defaults message,
-				timestamp: Date.now()
-				# Stop pubnub logging loads of "Missing Message" errors, as they are quite distracting
-				message: ' '
-
+		doPublish = ->
+			return if publishQueue[0].length is 0
+			message = publishQueue.shift()
 			pubnub.publish({ channel, message })
+			if publishQueue.length is 0
+				publishQueue = [[]]
+				publishQueueRemainingBytes = MAX_LOG_BYTE_SIZE
+			messageIndex = Math.max(messageIndex - 1, 0)
+			logsOverflow = false if messageIndex < MAX_MESSAGE_INDEX
+		setInterval(doPublish, LOG_PUBLISH_INTERVAL)
 
-		# Replay queue now that we have initialised the publish function
-		publish(args...) for args in publishQueue
+	return (message) ->
+		# Disable sending logs for bandwidth control
+		return if disableLogs or (messageIndex >= MAX_MESSAGE_INDEX and publishQueueRemainingBytes <= 0)
+		if _.isString(message)
+			message = { m: message }
 
-	return -> publishQueue.push(arguments)
+		_.defaults message,
+			t: Date.now()
+			m: ''
+		msgLength = Buffer.byteLength(JSON.stringify(message), 'utf8')
+		return if msgLength > MAX_LOG_BYTE_SIZE # Unlikely, but we can't allow this
+		remaining = publishQueueRemainingBytes - msgLength
+		if remaining >= 0
+			publishQueue[messageIndex].push(message)
+			publishQueueRemainingBytes = remaining
+		else if messageIndex < MAX_MESSAGE_INDEX
+			messageIndex += 1
+			publishQueue[messageIndex] = [ message ]
+			publishQueueRemainingBytes = MAX_LOG_BYTE_SIZE - msgLength
+		else if !logsOverflow
+			logsOverflow = true
+			messageIndex += 1
+			publishQueue[messageIndex] = [ { m: 'Warning! Some logs dropped due to high load', t: Date.now(), s: true } ]
```

Comment:
The in messageIndex is rather arbitrary, so that we limit the amount of RAM used for logs. This extra message will be sent in time by the doPublish interface, so it should be fine (albeit a bit suboptimal).